src/metabase/models/audit_log.clj

```diff
@@ -31,51 +34,73 @@
+        shared-updated-keys (set/intersection (set (keys previous-only)) (set (keys new-only)))]
+    {:previous (select-keys previous-object shared-updated-keys)
+     :new (select-keys object shared-updated-keys)}))
+
+(mu/defn record-event!
+  "Records an event in the Audit Log.
 
   `topic` is a keyword representing the type of event being recorded, e.g. `:dashboard-create`. If the keyword is
   namespaced (e.g. `:event/dashboard-create`) the namespace is stripped before the event is recorded.
 
-  `object` is typically the object that the event is acting on, e.g. a `Dashboard` instance. The details about the
-  object which are recorded are determined by the `model-details` multimethod, which is typically implemented in the
-  appropriate model namespace.
-
-  `object` can also be a map of arbitrary details relavent to the event, which is recorded as-is. If the name and/or ID
-  of a model are also relevant to the event and should be recorded, they can be passed as fourth and fifth arguments."
-  ([topic object]
-   (record-event! topic object api/*current-user-id*))
-
-  ([topic object user-id]
-   (record-event! topic object user-id (some-> (t2/model object) name)))
-
-  ([topic object user-id model]
-   (record-event! topic object user-id model (u/id object)))
+  `params` is a map that can optionally include the following fields:
+  - `:object`: the object the event is acting on, e.g. a `Card` instance
+  - `:previous-object`: the previous version of the object, for update events
+  - `:user-id`: the user ID that initiated the event (defaults: `api/*current-user-id*`)
+  - `:model`: the name of the model the event is acting on, e.g. `:model/Card` or \"Card\" (default: model of `:object`)
+  - `:model-id`: the ID of the model the event is acting on (default: ID of `:object`)
+  - `:details`: a map of arbitrary details relavent to the event, which is recorded as-is (default: {})
 
-  ([topic object user-id model model-id]
-   (let [unqualified-topic (keyword (name topic))
-         model-name        (model-name model)
-         details           (cond
-                             (:raw? object) (dissoc object :raw?)
-                             (t2/model object) (model-details object unqualified-topic)
-                             :else (or object {}))]
-     (t2/insert! :model/AuditLog
-                 :topic    unqualified-topic
-                 :details  details
-                 :model    model-name
-                 :model_id model-id
-                 :user_id  user-id)
-     ;; TODO: temporarily double-writing to the `activity` table, delete this in Metabase v48
-     (when-not (#{:card-read :dashboard-read :table-read :card-query :setting-update} unqualified-topic)
+  `:object` and `:previous-object` both have `model-details` called on them to determine which fields should be audited,
+  then they are added to `:details` before the event is recorded. `:previous-object` is only included if any audited fields
+  were updated."
```

Comment:
Here's the biggest change of the file. Reworked this function to have a cleaner interface, with a `params` map of entirely optional keys.